tests/RobotFramework/tests/tedge_agent/workflows/custom_operation.robot

```diff
@@ -98,6 +98,14 @@ Invoke sub-operation from a sub-operation
     ${expected_log}    Get File    ${CURDIR}/gp-command-expected.log
     Should Be Equal    ${actual_log}    ${expected_log}
 
+Command steps should not be executed twice
+    Execute Command    tedge mqtt pub --retain te/device/main///cmd/issue-2896/test-1 '{"status":"init"}'
+    ${cmd_messages}    Should Have MQTT Messages    te/device/main///cmd/issue-2896/test-1    message_pattern=.*successful.*
+    Execute Command    tedge mqtt pub --retain te/device/main///cmd/issue-2896/test-1 ''
+    ${workflow_log}    Execute Command    grep '\\[ issue-2896' /var/log/tedge/agent/workflow-issue-2896-test-1.log |cut -d '|' -f 1 | sort | sed 's/ $//'
```

Comment:
I'm not happy with this check which `sort`s the sequence of steps before comparison with a reference sequence.

I came with this because *two* sequences are valid (depending on which state is resumed first on restart (`issue-2896 > restart @ executing` or `issue-2896 @ await_restart`):

```
----------------------[ issue-2896 @ init 
----------------------[ issue-2896 @ executing 
----------------------[ issue-2896 @ await_restart 
----------------------[ issue-2896 > restart @ init 
----------------------[ issue-2896 > restart @ scheduled 
----------------------[ issue-2896 > restart @ executing 
----------------------[ issue-2896 > restart @ executing 
----------------------[ issue-2896 @ await_restart 
----------------------[ issue-2896 > restart @ successful 
----------------------[ issue-2896 @ await_restart 
----------------------[ issue-2896 @ restarted 
----------------------[ issue-2896 @ successful
```

as well as

```
----------------------[ issue-2896 @ init
----------------------[ issue-2896 @ executing
----------------------[ issue-2896 @ await_restart
----------------------[ issue-2896 > restart @ init
----------------------[ issue-2896 > restart @ scheduled
----------------------[ issue-2896 > restart @ executing
----------------------[ issue-2896 @ await_restart
----------------------[ issue-2896 > restart @ executing
----------------------[ issue-2896 > restart @ successful
----------------------[ issue-2896 @ await_restart
----------------------[ issue-2896 @ restarted
----------------------[ issue-2896 @ successful
``` 

Any idea for an alternative solution?

Reply:
Maybe something like: 

```
*** Variables ***
${VALID_SEQUENCE_1}   ${CURDIR}/valid_sequence_1.log
${VALID_SEQUENCE_2}   ${CURDIR}/valid_sequence_2.log

*** Keywords ***
Validate Log Sequence
    [Arguments]    ${log}    ${valid_sequences}
    FOR    ${sequence}    IN    @{valid_sequences}
        ${match}=    Run Keyword And Return Status    Compare Log With Pattern    ${log}    ${sequence}
        Run Keyword If    ${match}    RETURN    Log Sequence Matches
    END
    Fail    Log sequence does not match any valid sequence

Compare Log With Pattern
    [Arguments]    ${log}    ${pattern_file}
    ${pattern}    Get File    ${pattern_file}
    Should Match Regexp    ${log}    ${pattern}

*** Test Cases ***
Test Log Sequence
    Execute Command    tedge mqtt pub --retain te/device/main///cmd/issue-2896/test-1 '{"status":"init"}'
    ${cmd_messages}    Should Have MQTT Messages    te/device/main///cmd/issue-2896/test-1    message_pattern=.*successful.*
    Execute Command    tedge mqtt pub --retain te/device/main///cmd/issue-2896/test-1 ''
    ${workflow_log}    Execute Command    grep '\\[ issue-2896' /var/log/tedge/agent/workflow-issue-2896-test-1.log | cut -d '|' -f 1 | sed 's/ $//'
    Validate Log Sequence    ${workflow_log}    ${VALID_SEQUENCE_1}    ${VALID_SEQUENCE_2}
```

- Define Valid Sequences:

Save the valid sequences as regular expressions in files (valid_sequence_1.log and valid_sequence_2.log). Each line in these files should match a possible step in the log. You can use wildcards and regular expressions to accommodate slight variations.

- Validate Log Sequence:

The keyword Validate Log Sequence iterates over all provided valid sequences and checks if the log matches any of them using the Compare Log With Pattern keyword.

- Compare Log With Pattern:

This keyword reads the pattern from a file and checks if the log matches this pattern using the Should Match Regexp keyword.

This approach ensures that you are comparing the sequences as they are without sorting, allowing for flexibility while maintaining correctness.

Reply:
I finally opted for a third solution, using `TRY / EXCEPT`:

```
 ${workflow_log}    Execute Command    grep '\\[ issue-2896' /var/log/tedge/agent/workflow-issue-2896-test-1.log | cut -d '|' -f 1 | sed 's/ $//'
    TRY
        # "issue-2896 @ await_restart" is resumed first
        ${expected_log}    Get File    ${CURDIR}/issue-2896-test-1.seq1.log
        Should Be Equal    ${workflow_log}    ${expected_log}
    EXCEPT
        # "issue-2896 > restart @ executing" is resumed first
        ${expected_log}    Get File    ${CURDIR}/issue-2896-test-1.seq2.log
        Should Be Equal    ${workflow_log}    ${expected_log}
    END
```